dynamiq/memory/backends/postgresql.py

```diff
@@ -0,0 +1,305 @@
+    """
+    _CLEAR_SQL: ClassVar[
+        str
+    ] = """
+    TRUNCATE TABLE "{table_name}";
+    """
+
+    @property
+    def to_dict_exclude_params(self) -> dict[str, bool]:
+        """Define parameters to exclude during serialization."""
+        return super().to_dict_exclude_params | {"_conn": True}
+
+    def to_dict(self, include_secure_params: bool = False, **kwargs) -> dict[str, Any]:
+        """Converts the instance to a dictionary."""
+        exclude = kwargs.pop("exclude", self.to_dict_exclude_params.copy())
+        data = self.model_dump(exclude=exclude, **kwargs)
+        data["connection"] = self.connection.to_dict(include_secure_params=include_secure_params)
+        if "type" not in data:
+            data["type"] = self.type
+        return data
+
+    def model_post_init(self, __context: Any) -> None:
+        """Initialize the PostgreSQL connection and ensure table exists."""
+        try:
+            self._conn = self.connection.connect()
+            if self.create_if_not_exist:
+                self._create_table_and_indices()
+            logger.debug(f"PostgreSQL backend connected to table '{self.table_name}'.")
+        except psycopg.Error as e:
+            logger.error(f"Failed to initialize PostgreSQL connection or table '{self.table_name}': {e}")
+            raise PostgresMemoryError(f"Failed to initialize PostgreSQL connection or table: {e}") from e
+        except Exception as e:
+            logger.error(f"Unexpected error initializing PostgreSQL backend: {e}")
+            raise PostgresMemoryError(f"Unexpected error initializing PostgreSQL backend: {e}") from e
+
+    def _execute_sql(self, sql: str, params: tuple | list | None = None, fetch: str | None = None):
+        """Helper to execute SQL, handling potential connection issues."""
+        if self._conn is None or self._conn.closed:
+            logger.warning("PostgreSQL connection lost or not initialized. Attempting to reconnect.")
+            try:
+                self._conn = self.connection.connect()
+            except Exception as e:
+                raise PostgresMemoryError(f"Failed to re-establish PostgreSQL connection: {e}") from e
+
+        try:
+            with self._conn.cursor() as cur:
+                cur.execute(sql, params)
+                if fetch == "one":
+                    return cur.fetchone()
+                elif fetch == "all":
+                    return cur.fetchall()
+                return None
+        except psycopg.Error as e:
+            logger.error(f"PostgreSQL error executing SQL: {e}\nSQL: {sql}\nParams: {params}")
+            raise PostgresMemoryError(f"PostgreSQL error: {e}") from e
+
+    def _create_table_and_indices(self) -> None:
+        """Creates the table and necessary indices if they don't exist."""
+        logger.debug(f"Ensuring table '{self.table_name}' and indices exist...")
+        table_sql = self._CREATE_TABLE_SQL.format(
```

Comment:
Such SQL formatting/execution is unsafe, please check `pgvector.py` storage and define/format SQL queries in the same way (use SQL(<>) and proper format)